ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java

```diff
@@ -0,0 +1,429 @@
+    for (int i = 0; i < mockServers.size(); i++) {
+      MockServer mockServer = mockServers.get(i);
+      if (mockServer.getDataCenter().equals(routerConfig.routerDatacenterName)) {
+        mockServer.setServerErrorForAllRequests(ServerErrorCode.No_Error);
+        break;
+      }
+    }
+
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      List<ResponseInfo> responses = sendAndWaitForResponses(requestListToFill);
+      for (ResponseInfo response : responses) {
+        op.handleResponse(response);
+        if (op.isOperationComplete()) {
+          break;
+        }
+      }
+    }
+
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    assertSuccess(op);
+  }
+
+  /**
+   * Submit all the requests that were handed over by the operation and wait until a response is received for every
+   * one of them.
+   * @param requestList the list containing the requests handed over by the operation.
+   * @return the list of responses from the network client.
+   * @throws IOException
+   */
+  private List<ResponseInfo> sendAndWaitForResponses(List<RequestInfo> requestList)
+      throws IOException {
+    List<ResponseInfo> responseList = new ArrayList<>();
+    int sendCount = requestList.size();
+    responseList.addAll(networkClient.sendAndPoll(requestList));
+    requestList.clear();
+    while (responseList.size() < sendCount) {
+      responseList.addAll(networkClient.sendAndPoll(requestList));
+    }
+    return responseList;
+  }
+
+  /**
+   * Assert that the operation is complete and successful. Note that the future completion and callback invocation
+   * happens outside of the GetOperation, so those are not checked here. But at this point, the operation result should
+   * be ready.
+   * @param op the {@link GetBlobInfoOperation} that should have completed.
+   */
+  private void assertSuccess(GetBlobInfoOperation op) {
+    Assert.assertEquals("Null expected", null, op.getOperationException());
+    BlobInfo blobInfo = op.getOperationResult();
+    Assert.assertTrue("Blob properties must be the same",
+        RouterTestHelpers.haveEquivalentFields(blobProperties, blobInfo.getBlobProperties()));
+    Assert.assertArrayEquals("User metadata must be the same", userMetadata, blobInfo.getUserMetadata());
+  }
+
+  /**
+   * Get the properties for the {@link NonBlockingRouter}.
+   * @return the constructed properties.
+   */
+  private VerifiableProperties getNonBlockingRouterProperties() {
+    Properties properties = new Properties();
+    properties.setProperty("router.hostname", "localhost");
+    properties.setProperty("router.datacenter.name", "DC1");
+    properties.setProperty("router.get.request.parallelism", Integer.toString(requestParallelism));
+    properties.setProperty("router.get.success.target", Integer.toString(successTarget));
+    return new VerifiableProperties(properties);
+  }
+}
```

Comment:
These tests cover a lot of cases, but I do think there might be some that we have either missed out or discover as we develop other parts. What do you think about starting a google doc with test scenarios listed and marking the ones that are already covered ? (for all operations). When we eventually come to the point where we have to test correctness and list test cases to try out (or write integration tests for), we would already know what has to be tested (before any context fades from memory) and we would have confidence in most scenarios.

This might be especially true with respect to the mixture of different types of responses from the servers (and related to our previous discussion on creating a generic test framework for this). This applies to all operations.

At the end of development, I really think it is worth the time investing in building a simple framework that tests response mixtures and it will give us a lot of confidence w.r.t correctness.